### PR TITLE
[WIP] initial work

### PR DESCRIPTION
**Pull request by knowbody**
#368

### CODE REVIEW

File: src/createStore.js [hunk 1]
```diff
@@ -92,10 +93,20 @@ export default function createStore(reducer, initialState) {
...
+    );
+
+    isDispatching = true;
 
-    currentState = currentReducer(currentState, action);
-    listeners.forEach(listener => listener());
-    return action;
+    try {
+      currentState = currentReducer(currentState, action);
+      listeners.forEach(listener => listener());
```
Comment (gaearon) [inline src/createStore.js]: We want to move this line out. It's OK to call the listeners. The only line we want to guard is `currentState` assignment.

File: src/createStore.js [hunk 2]
```diff
@@ -92,10 +93,20 @@ export default function createStore(reducer, initialState) {
       isPlainObject(action),
       'Actions must be plain objects. Use custom middleware for async actions.'
     );
+    invariant(
+      !isDispatching,
+      'Cannot dispatch in the middle of a dispatch.'
```
Comment (gaearon) [inline src/createStore.js]: Maybe change the wording `Cannot dispatch in the middle of a reducer call.`? To clarify this isn't exactly the same thing as what Flux has.

Comment (knowbody) [inline src/createStore.js]: sure, adding those changes and testing. thanks @gaearon

File: src/createStore.js (hunk 1 above)
Comment (knowbody) [inline src/createStore.js]: so you would notify listeners before `try` of in the `finally`?

**Comment by knowbody** [comment]
@gaearon this can be closed as it's done here #372

File: src/createStore.js [hunk 3]
```diff
@@ -92,8 +93,19 @@ export default function createStore(reducer, initialState) {
...
       'Actions must be plain objects. Use custom middleware for async actions.'
     );
+    invariant(
+      !isDispatching,
+      'Cannot dispatch in the middle of reducer call.'
+    );
+
+    isDispatching = true;
+
+    try {
```
Comment (taylorhakes) [inline src/createStore.js]: We can't add a `try {} catch {}` here. This will cause way more problems then it solves. All reducer errors will be swallowed.

**Comment by gaearon** [comment]
Closing in favor of #372, thanks for your effort!

**Pull request closed without merging**